### PR TITLE
[compiler] extend + fix simplifier for integral types

### PR DESCRIPTION
**Pull request by ehigham**
Fix simplification for integer subtraction
Extend scope of simplifier to more numeric types.

### CODE REVIEW

**Comment by ehigham** [comment]
```
ail version: 0.2.109-c163bcb21073
Error summary: AssertionError: assertion failed
self = <test.hail.linalg.test_linalg.Tests testMethod=test_tree_matmul>

    @fails_service_backend()
    @fails_local_backend()
    def test_tree_matmul(self):
        nm = np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
        m = BlockMatrix.from_numpy(nm, block_size=2)
        nrow = np.array([[7.0, 8.0, 9.0]])
        row = BlockMatrix.from_numpy(nrow, block_size=2)
    
>       with BatchedAsserts() as b:

test/hail/linalg/test_linalg.py:612: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/hail/linalg/test_linalg.py:96: in __exit__
    vals.extend(list(hl.eval(tuple([all_bms[k] for k in bm_keys[batch_start:batch_start + batch_size]]))))
<decorator-gen-692>:2: in eval
    ???
hail/typecheck/check.py:577: in wrapper
    return __original_func(*args_, **kwargs_)
hail/expr/expressions/expression_utils.py:191: in eval
    return eval_timed(expression)[0]
<decorator-gen-690>:2: in eval_timed
    ???
hail/typecheck/check.py:577: in wrapper
    return __original_func(*args_, **kwargs_)
hail/expr/expressions/expression_utils.py:161: in eval_timed
    return Env.backend().execute(MakeTuple([ir]), timed=True)[0]
hail/backend/py4j_backend.py:82: in execute
    raise e.maybe_user_error(ir) from None
hail/backend/py4j_backend.py:76: in execute
    result_tuple = self._jbackend.executeEncode(jir, stream_codec, timed)
../../.venv/lib/python3.10/site-packages/py4j/java_gateway.py:1321: in __call__
    return_value = get_return_value(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = ('xro549', <py4j.clientserver.JavaClient object at 0x7fd0d58f6fb0>, 'o1', 'executeEncode')
kwargs = {}
pyspark = <module 'pyspark' from '/home/edmund/.local/src/hail/.venv/lib/python3.10/site-packages/pyspark/__init__.py'>
s = 'java.lang.AssertionError: assertion failed', tpl = JavaObject id=o550
deepest = 'AssertionError: assertion failed'
full = 'java.lang.AssertionError: assertion failed\n\tat scala.Predef$.assert(Predef.scala:208)\n\tat is.hail.expr.ir.BlockMa...lientServerConnection.run(ClientServerConnection.java:106)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\n\n\n'
error_id = -1

    def deco(*args, **kwargs):
        import pyspark
        try:
            return f(*args, **kwargs)
        except py4j.protocol.Py4JJavaError as e:
            s = e.java_exception.toString()
    
            # py4j catches NoSuchElementExceptions to stop array iteration
            if s.startswith('java.util.NoSuchElementException'):
                raise
    
            tpl = Env.jutils().handleForPython(e.java_exception)
            deepest, full, error_id = tpl._1(), tpl._2(), tpl._3()
>           raise fatal_error_from_java_error_triplet(deepest, full, error_id) from None
E           hail.utils.java.FatalError: AssertionError: assertion failed
E           
E           Java stack trace:
E           java.lang.AssertionError: assertion failed
E           	at scala.Predef$.assert(Predef.scala:208)
E           	at is.hail.expr.ir.BlockMatrixMap.execute(BlockMatrixIR.scala:269)
E           	at is.hail.expr.ir.BlockMatrixMap2.execute(BlockMatrixIR.scala:393)
E           	at is.hail.expr.ir.Interpret$.run(Interpret.scala:875)
E           	at is.hail.expr.ir.Interpret$.alreadyLowered(Interpret.scala:59)
E           	at is.hail.expr.ir.LowerOrInterpretNonCompilable$.evaluate$1(LowerOrInterpretNonCompilable.scala:20)
E           	at is.hail.expr.ir.LowerOrInterpretNonCompilable$.rewrite$1(LowerOrInterpretNonCompilable.scala:67)
E           	at is.hail.expr.ir.LowerOrInterpretNonCompilable$.rewrite$1(LowerOrInterpretNonCompilable.scala:53)
E           	at is.hail.expr.ir.LowerOrInterpretNonCompilable$.apply(LowerOrInterpretNonCompilable.scala:72)
E           	at is.hail.expr.ir.lowering.LowerOrInterpretNonCompilablePass$.transform(LoweringPass.scala:69)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$3(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$1(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply(LoweringPass.scala:14)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply$(LoweringPass.scala:13)
E           	at is.hail.expr.ir.lowering.LowerOrInterpretNonCompilablePass$.apply(LoweringPass.scala:64)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1(LoweringPipeline.scala:22)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1$adapted(LoweringPipeline.scala:20)
E           	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
E           	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
E           	at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:38)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.apply(LoweringPipeline.scala:20)
E           	at is.hail.expr.ir.CompileAndEvaluate$._apply(CompileAndEvaluate.scala:47)
E           	at is.hail.backend.spark.SparkBackend._execute(SparkBackend.scala:454)
E           	at is.hail.backend.spark.SparkBackend.$anonfun$executeEncode$2(SparkBackend.scala:490)
E           	at is.hail.backend.ExecuteContext$.$anonfun$scoped$3(ExecuteContext.scala:75)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.backend.ExecuteContext$.$anonfun$scoped$2(ExecuteContext.scala:75)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.annotations.RegionPool$.scoped(RegionPool.scala:17)
E           	at is.hail.backend.ExecuteContext$.scoped(ExecuteContext.scala:63)
E           	at is.hail.backend.spark.SparkBackend.withExecuteContext(SparkBackend.scala:342)
E           	at is.hail.backend.spark.SparkBackend.$anonfun$executeEncode$1(SparkBackend.scala:487)
E           	at is.hail.utils.ExecutionTimer$.time(ExecutionTimer.scala:52)
E           	at is.hail.backend.spark.SparkBackend.executeEncode(SparkBackend.scala:486)
E           	at jdk.internal.reflect.GeneratedMethodAccessor72.invoke(Unknown Source)
E           	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
E           	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
E           	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
E           	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
E           	at py4j.Gateway.invoke(Gateway.java:282)
E           	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
E           	at py4j.commands.CallCommand.execute(CallCommand.java:79)
E           	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
E           	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
E           	at java.base/java.lang.Thread.run(Thread.java:834)
E           
E           
E           
E           Hail version: 0.2.109-c163bcb21073
E           Error summary: AssertionError: assertion failed

hail/backend/py4j_backend.py:35: FatalError

```

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala [hunk 1]
```diff
@@ -146,6 +148,111 @@ object Simplify {
...
+        }
+      case _ =>
+        None
+    }
+
+    Array(
+      hoistUnaryOp,
+      (ir: IR) => integralBinaryIdentities(Literal.coerce(ir.typ, _))(ir),
+      (ir: IR) => commonBinaryIdentities(Literal.coerce(ir.typ, _))(ir),
+    ).reduce((f, g) =>
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala:248]: It's not clear to me that this does any less allocation than `orElse`. I agree we should be conscious of excessive allocation, but in this case I don't see an issue with `orElse`. (And we're allocating thunks all over the place in simplify already.)

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala [hunk 2]
```diff
@@ -1012,5 +1114,7 @@ object Simplify {
       val needsDense = sparsityStrategy == NeedsDense || sparsityStrategy.exists(leftBlock = false, rightBlock = true)
       val maybeDense = if (needsDense) BlockMatrixDensify(left) else left
       BlockMatrixMap(maybeDense, leftName, Subst(f, BindingEnv.eval(rightName -> getElement)), needsDense)
+    case BlockMatrixMap(matrix, name, Ref(x, _), _)  if name == x =>
+      matrix
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/Simplify.scala:1118]: The new simplify rules could also create a map with a constant body. We could catch those cases with something like
```
  case BlockMatrixMap(matrix, name, const@(Ref(x, _) if name != x | _ if IsConstant(const)), _) =>
    val typ = matrix.typ
    BlockMatrixBroadcast(
      ValueToBlockMatrix(const, FastIndexedSeq(1, 1), typ.blockSize),
      FastIndexedSeq(),
      typ.shape,
      typ.blockSize)
```

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
great addition!

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala:248]: Thanks!

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala (hunk 2 above)
Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala:1118]: Oh good idea. Thank you :)

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala [hunk 3]
```diff
@@ -146,6 +148,111 @@ object Simplify {
...
+            if (x == pure(1)) Some(y)
+            else if (x == pure(-1)) Some(ApplyUnaryPrimOp(Negate(), y))
+            else if (y == pure(1)) Some(x)
+            else if (y == pure(-1)) Some(ApplyUnaryPrimOp(Negate(), x))
+            else None
+
+          case RoundToNegInfDivide() =>
+            if (y == pure(1)) Some(x)
+            else if (y == pure(-1)) Some(ApplyUnaryPrimOp(Negate(), x))
+            else if (y == pure(0)) Some(Die("division by zero", ir.typ))
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Simplify.scala:234]: this doesn't error, it returns one of (-Inf, Inf, NaN)

File: hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala [hunk 4]
```diff
@@ -266,7 +266,13 @@ case class BlockMatrixMap(child: BlockMatrixIR, eltName: String, f: IR, needsDen
     f(_, scalar)
 
   override protected[ir] def execute(ctx: ExecuteContext): BlockMatrix = {
-    assert(f.isInstanceOf[ApplyUnaryPrimOp] || f.isInstanceOf[Apply] || f.isInstanceOf[ApplyBinaryPrimOp])
+    assert(
+      f.isInstanceOf[ApplyUnaryPrimOp]
+        || f.isInstanceOf[Apply]
+        || f.isInstanceOf[ApplyBinaryPrimOp]
+        || f.isInstanceOf[Ref]
```
[the remark this answers is not in the document]
Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala:273]: Thank you - I had missed those!

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala (hunk 3 above)
Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala:234]: Interesting... We allow division by zero?

Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala:234]: Even in the case of ints?

Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala:234]: No, I get
```
>> hl.eval(hl.int32(5) // hl.int32(0))
...
java.lang.ArithmeticException: / by zero
	at java.lang.Math.floorDiv(Math.java:1052)
...
```

File: hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala:273]: I think
```
case Ref(`eltName`, _) => ("id", x => x)
```
should work

[the remark this answers is not in the document]
Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala:273]: thanks :)

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala (hunk 3 above)
Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala:234]: That's super helpful. Thank you.

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**